Move from CSP-style format to JSON Field Values

juliatuttle opened the pull request:
I'd like to bring this in line with Reporting in using JSON Field Values instead of CSP-style header values, mostly because I believe JSON parsers (or at least Chromium's) are more well-tested than one-off parsers written for CSP-style values.

igrigorik commented on index.html:
```diff
@@ -127,63 +127,42 @@
-directive                 = directive-name [ "=" directive-value ]
-directive-name            = token
-directive-value           = token | quoted-string
+NEL = <a>json-field</a>
```
nitpick: `json-field-value`

igrigorik commented on index.html:
```diff
@@ -127,63 +127,42 @@
         <section>
-          <h2>The `report-uri` Directive</h2>
-          <p>The <dfn>report-uri</dfn> directive specifies a URL to which the user agent sends reports about network errors. The <a>report-uri</a> directive is a REQUIRED field to register an <a>NEL policy</a>, and OPTIONAL if the intent is to remove a previous registration - see <a>max-age</a>. The ABNF grammar for the name and value of the directive is:</p>
+          <h2>The `report-uri` Field</h2>
```
I think we should remove report-uri entirely, to make everyone's life easier; we should assume that Reporting API exists, and use it. Practically speaking, that means the site has to specify two headers:

> Report-To: { "url": "https://example.com/reports", "group": "log-service", "max-age": 10886400}
> NEL: { "report-to": "log-service", "includeSubdomains": true, "max-age": 86400}

- We should delegate all the endpoint registration and handling to Reporting. 
- Should we allow multiple NEL policies? E.g. the "failover" use case?
- There's some interesting complications here with max-age.. 
  - What happens when report-to group's lifetime expires, while the NEL policy is still active? 
  - We could drop max-age on NEL policy and rely on Report-To group's lifecycle. However, this makes it tricky to reason about cases where you have a single report-to group associated with multiple services (e.g. NEL and CSP reports). Perhaps two separate lifetimes is OK; we just need to think through the possible workflows.

juliatuttle commented:
I agree we should move to report-to; I was going to do that in a separate
commit, because it feels cleaner to me, and in case there is substantial
discussion/pushback on either.

I agree on the endpoint registration -- NEL will just remember the group
name, expiration time, and subdomains flag.

I don't think failover makes sense for NEL -- Reporting doesn't provide a
"this report upload failed" callback, since it provides its own retry and
failover internally. We could allow multiple *parallel* reports to go to
different endpoint groups, if there was demand. As an example, this would
allow sites to feed both their own monitoring and an external service.



On Fri, Aug 4, 2017 at 7:59 PM Ilya Grigorik <notifications@github.com>
wrote:

> *@igrigorik* commented on this pull request.
>
> Can we drop report-uri in favor of report-to? Now that Reporting API is a
> thing...
> ------------------------------
>
> In index.html
> <https://github.com/WICG/network-error-logging/pull/61#discussion_r131506093>
> :
>
> > @@ -127,63 +127,42 @@
>          <p>The <dfn>NEL header field</dfn> is used to communicate the <dfn>NEL policy</dfn> to the user agent. The ABNF (Augmented Backus-Naur Form) syntax for the <a>NEL header field</a> is as follows:</p>
>
>          <pre>
> -NEL = "NEL" ":" [ directive ]  \\\*( ";" [ directive ] )
> -
> -directive                 = directive-name [ "=" directive-value ]
> -directive-name            = token
> -directive-value           = token | quoted-string
> +NEL = <a>json-field</a>
>
> nitpick: json-field-value
> ------------------------------
>
> In index.html
> <https://github.com/WICG/network-error-logging/pull/61#discussion_r131507261>
> :
>
> >
>          <p>The user agent MUST ignore the NEL header specified via a <code>meta</code> element to mitigate hijacking of error reporting via scripting attacks. The <a>NEL policy</a> MUST be delivered via the <a>NEL header field</a>.</p>
>
>          <p class="note">The restriction on <code>meta</code> element is consistent with [[CSP]] specification, which restricts reporting registration to HTTP header fields only for same reasons.</p>
>
>          <section>
> -          <h2>The `report-uri` Directive</h2>
> -          <p>The <dfn>report-uri</dfn> directive specifies a URL to which the user agent sends reports about network errors. The <a>report-uri</a> directive is a REQUIRED field to register an <a>NEL policy</a>, and OPTIONAL if the intent is to remove a previous registration - see <a>max-age</a>. The ABNF grammar for the name and value of the directive is:</p>
> +          <h2>The `report-uri` Field</h2>
>
> I think we should remove report-uri entirely, to make everyone's life
> easier; we should assume that Reporting API exists, and use it. Practically
> speaking, that means the site has to specify two headers:
>
> Report-To: { "url": "https://example.com/reports", "group":
> "log-service", "max-age": 10886400}
> NEL: { "report-to": "log-service", "includeSubdomains": true, "max-age":
> 86400}
>
>
>    - We should delegate all the endpoint registration and handling to
>    Reporting.
>    - Should we allow multiple NEL policies? E.g. the "failover" use case?
>    - There's some interesting complications here with max-age..
>       - What happens when report-to group's lifetime expires, while the
>       NEL policy is still active?
>       - We could drop max-age on NEL policy and rely on Report-To group's
>       lifecycle. However, this makes it tricky to reason about cases where you
>       have a single report-to group associated with multiple services (e.g. NEL
>       and CSP reports). Perhaps two separate lifetimes is OK; we just need to
>       think through the possible workflows.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/WICG/network-error-logging/pull/61#pullrequestreview-54499207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYoj0bL5dFUmYGak7aIEdNXEwa5S3gFks5sU7BjgaJpZM4Ot9SJ>
> .
>

igrigorik commented:
> I agree we should move to report-to; I was going to do that in a separate
> commit, because it feels cleaner to me, and in case there is substantial
> discussion/pushback on either.

Honestly, I don't see any reason why we should bother with report-uri; I guess I'm pushing back on the unnecessary (I think) intermediate step. Let's jump to report-to? :)

> I agree on the endpoint registration -- NEL will just remember the group
> name, expiration time, and subdomains flag.

Sounds good. What's your thinking on the mismatched expire's I mentioned above? 

> I don't think failover makes sense for NEL -- Reporting doesn't provide a
> "this report upload failed" callback, since it provides its own retry and
> failover internally. We could allow multiple *parallel* reports to go to
> different endpoint groups, if there was demand. As an example, this would
> allow sites to feed both their own monitoring and an external service.

Right, makes sense.

juliatuttle commented:
I'm not worried about the mismatched expires. Maybe we should add a
clarifying note explaining that site owners need to keep both of those
current. (Honestly, I'm expecting sites will set really high max-ages for
Reporting and lower ones for the individual reporting features, so they can
dial them up or down as needed.)

I'll change it to an array of endpoint groups, then, and do both in a
single PR?



On Sat, Aug 5, 2017 at 11:57 AM Ilya Grigorik <notifications@github.com>
wrote:

> I agree we should move to report-to; I was going to do that in a separate
> commit, because it feels cleaner to me, and in case there is substantial
> discussion/pushback on either.
>
> Honestly, I don't see any reason why we should bother with report-uri; I
> guess I'm pushing back on the unnecessary (I think) intermediate step.
> Let's jump to report-to? :)
>
> I agree on the endpoint registration -- NEL will just remember the group
> name, expiration time, and subdomains flag.
>
> Sounds good. What's your thinking on the mismatched expire's I mentioned
> above?
>
> I don't think failover makes sense for NEL -- Reporting doesn't provide a
> "this report upload failed" callback, since it provides its own retry and
>
> failover internally. We could allow multiple *parallel* reports to go to
>
>
> different endpoint groups, if there was demand. As an example, this would
> allow sites to feed both their own monitoring and an external service.
>
> Right, makes sense.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/WICG/network-error-logging/pull/61#issuecomment-320451391>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYoj43A6fWSgrMI42YQwGwRwhP_YBAYks5sVJDLgaJpZM4Ot9SJ>
> .
>

igrigorik commented:
> I'm not worried about the mismatched expires. Maybe we should add a
> clarifying note explaining that site owners need to keep both of those
> current. (Honestly, I'm expecting sites will set really high max-ages for
> Reporting and lower ones for the individual reporting features, so they can
> dial them up or down as needed.)

Yep, makes sense. We should definitely add a note for this.

> I'll change it to an array of endpoint groups, then, and do both in a
> single PR?

Do you mean `report-to: [a, b, c]`? Before we jump into that.. A single group can have [multiple endpoints](http://wicg.github.io/reporting/#examples) associated with it. What do we "win" by adding multiple report groups here? I'm guess I'm looking for a crisp use case...

juliatuttle commented:
Hey @igrigorik, can you take a look at my new changes?

I abandoned the fan-out idea; multiple NEL headers (or one with multiple comma-separated dictionaries) are parsed as "first valid config wins".

I switched to Reporting, and dropped the age field since it's included in the metadata.

I added notes that max-age and includeSubdomains need to be replicated in Reporting for proper delivery.

I also corrected a few random errors I found, including my own name (!).

igrigorik commented on index.html:
```diff
@@ -73,24 +78,28 @@
     <p>Today, application developers do not have real-time web application availability data from their end users. For example, if the user fails to load the page due to a network error, such as a failed DNS lookup, a connection timeout, a reset connection, or other reasons, the site developer is unable to detect and address this issue. Existing methods, such as synthetic monitoring provide a par ...
 
-    <p>Network Error Logging (<dfn>NEL</dfn>) addresses this need by defining a mechanism enabling web applications to declare a reporting policy that can be used by the user agent to report network errors for a given origin. To take advantage of <a>NEL</a>, a web application opts into using <a>NEL</a> by supplying a <code>NEL</code> HTTP response header field that describes the reporting policy. ...
+    <p>Network Error Logging (<dfn>NEL</dfn>) addresses this need by defining a mechanism enabling web applications to declare a reporting policy that can be used by the user agent to report network errors for a given origin. To take advantage of <a>NEL</a>, a web application opts into using <a>NEL</a> by supplying a <code>NEL</code> HTTP response header field that describes the reporting policy. ...
```
Instead of `<a href="https://wicg.github.io/reporting/">Reporting API</a>`

Use: Reporting API [[!REPORTING]].

igrigorik commented on index.html:
```diff
@@ -104,10 +113,13 @@
         <dt><dfn>URL</dfn></dt>
         <dd>Defined by [[URL]].</dd>
+        <dt><dfn>endpoint group</dfn></dt>
+        <dd>Defined in the <a href="https://wicg.github.io/reporting/#id-member">Reporting API specification</a>.</dd>
```
`<a data-cite="REPORTING-1#id-member">` should generate the right link.

igrigorik commented on index.html:
```diff
@@ -120,70 +132,48 @@
-          <h2>The `includeSubDomains` Directive</h2>
-          <p>The OPTIONAL <dfn>includeSubDomains</dfn> directive is a valueless directive that, if present, signals the user agent that the <a>NEL policy</a> applies not only to the <a href="https://tools.ietf.org/html/rfc6454#section-3.2">origin</a> that served the <a href="https://tools.ietf.org/html/rfc7231#section-3">resource representation</a>, but also to any <a href="https://tools.ietf.org ...
+          <h2>The `includeSubDomains` Field</h2>
+          <p>The OPTIONAL <dfn>includeSubDomains</dfn> field, if present and true, signals the user agent that the <a>NEL policy</a> applies not only to the <a href="https://tools.ietf.org/html/rfc6454#section-3.2">origin</a> that served the <a href="https://tools.ietf.org/html/rfc7231#section-3">resource representation</a>, but also to any <a href="https://tools.ietf.org/html/rfc6454#section-3.2 ...
```
From an authoring perspective, we want to define the concept once and then let respec create the links for us:

- we should add a dependencies section that defines external concepts
- then just use `<a>concept</a>` to autogenerate correct links

Example: https://github.com/w3c/beacon/blob/gh-pages/index.html#L191-L312

juliatuttle commented:
I'll leave those changes for another PR, since they're more general structural changes (or dependent thereupon).
